File: src/chain-observers/providers/feed-price-provider.ts
import { price } from "@hiveio/wax";
import { WorkerBeeIterable } from "../../types/iterator";
import { FeedPriceClassifier } from "../classifiers";
import { TRegisterEvaluationContext } from "../classifiers/collector-classifier-base";
import { TProviderEvaluationContext } from "../factories/data-evaluation-context";
import { ProviderBase } from "./provider-base";

export interface IFeedPriceData {
  priceHistory: WorkerBeeIterable<price>;
  currentMinHistory: price;
  currentMaxHistory: price;
  currentMedianHistory: price;
}

export interface IFeedPriceProviderData {
  feedPrice?: IFeedPriceData;
};

export class FeedPriceProvider extends ProviderBase<{}, IFeedPriceProviderData> {
  public usedContexts(): Array<TRegisterEvaluationContext> {
    return [
      FeedPriceClassifier
    ]
  }

  public get baseStructure(): IFeedPriceProviderData {
    return {};
  }

  public async provide(data: TProviderEvaluationContext): Promise<IFeedPriceProviderData> {
    const result = this.baseStructure;

    const { currentMedianHistory, currentMinHistory, currentMaxHistory, priceHistory } = await data.get(FeedPriceClassifier);

    result.feedPrice = {
      currentMedianHistory,
      currentMinHistory,
      currentMaxHistory,
      priceHistory: new WorkerBeeIterable(priceHistory)
    };

    return result;
  }
}

File: python/workerbee/chain_observers/providers/feed_price_provider.py
"""FeedPriceProvider — provides feed price data.

Mirrors src/chain-observers/providers/feed-price-provider.ts.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from ..classifiers.collector_classifier_base import TRegisterEvaluationContext
from ..classifiers.feed_price_classifier import FeedPriceClassifier
from .provider_base import ProviderBase

if TYPE_CHECKING:
    from ..factories.data_evaluation_context import DataEvaluationContext
    from ..payloads import FeedPricePayload


class FeedPriceProvider(ProviderBase):
    def used_contexts(self) -> list[TRegisterEvaluationContext]:
        return [FeedPriceClassifier]

    async def provide(self, data: DataEvaluationContext) -> FeedPricePayload:
        feed_price_data = await data.get(FeedPriceClassifier)

        return {
            "feed_price": {
                "current_median_history": feed_price_data["current_median_history"],
                "current_min_history": feed_price_data["current_min_history"],
                "current_max_history": feed_price_data["current_max_history"],
                "price_history": feed_price_data["price_history"],
            },
        }
